modules/internal/ChapelIO.chpl

```diff
@@ -17,36 +17,310 @@
+      data += s.substring(1);
+    }
+
+    // writeThis will be called when
+    // e.g. stdout.write(c) is called
+    proc writeThis(x: Writer) {
+      x.write(data);
+    }
+  }
+
+  var c = new C();
+  // c.write will invoke c.writePrimitive
+  // with each numeric value in turn
+  c.write(41, 32, 23, 14);
+  // writeln is stdout.writeln and will invoke
+  // c.writeThis.
+  writeln(c);
+  // prints out:
+  // 4321
+
+  delete c;
+
+
+.. _default-write-and-read-methods:
+
+Default write and read Methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Default ``write`` methods are created for all types for which a user-defined
+``write`` method is not provided.  They have the following semantics:
+
+* for an array argument, outputs the elements of the array in row-major order
+  where rows are separated by line-feeds and blank lines are used to separate
+  other dimensions.
+* for a `domain` argument,  outputs the dimensions of the domain enclosed by
+  ``[`` and ``]``.
+* for a `range` argument, output the lower bound of the range followed by
+  ``..`` followed by the upper bound of the range.  If the stride of the range
+  is not one, the output is additionally followed by the word ``by`` followed
+  by the stride of the range.
+* for a tuples, outputs the components of the tuple in order delimited by ``(``
+  and ``)``, and separated by commas.
+* for a class, outputs the values within the fields of the class prefixed by
+  the name of the field and the character ``=``.  Each field is separated by a
+  comma.  The output is delimited by ``{`` and ``}``.
+* for a record, outputs the values within the fields of the class prefixed by
+  the name of the field and the character ``=``.  Each field is separated by a
+  comma.  The output is delimited by ``(`` and ``)``.
+
+Default ``read`` methods are created for all types for which a user-defined
+``read`` method is not provided.  The default ``read`` methods are defined to
+read in the output of the default ``write`` method.
+
+.. note::
+
+  Note that it is not currently possible to read and write circular
+  data structures with these mechanisms.
+
+ */
 module ChapelIO {
   use ChapelBase; // for uint().
   use SysBasic;
   // use IO; happens below once we need it.
   
+  // TODO -- this should probably be private
+  pragma "no doc"
   proc _isNilObject(val) {
     proc helper(o: object) return o == nil;
     proc helper(o)         return false;
     return helper(val);
   }
   
+  /*
+     An abstract base class type for classes that know how
+     to write values.
+   */
   class Writer {
+    /* returns `true` */
     proc writing param return true;
-    // if it's binary, we don't decorate class/record fields and values
+    /* returns `true` if this Writer is configured for binary I/O. */
     proc binary():bool { return false; }
-    // get other style elements
+    /* return other style elements.
+       subclasses should override this method */
+    pragma "no doc"
```

Comment:
Yes, we should, but it exists because Writer != channel and I want to merge them. But it's a documentation challenge because the arguments you have to provide to styleElement are all things like QIO_STYLE_ELEMENT_AGGREGATE and these are defined it ChapelBase. So, we could document it here, but I'd have to repeat the constants you could use but I'm too lazy...